pkg/lib/catalogsource/image_template.go

```diff
@@ -0,0 +1,356 @@
+
+	// templ_gvk is a dynamic template that uses its own subgroups
+	templ_gvk = "{group:(?P<group>.*?),version:(?P<version>.*?),kind:(?P<kind>.*?),name:(?P<name>.*?),namespace:(?P<namespace>.*?),jsonpath:(?P<jsonpath>{.*?})}"
+
+	// templateMissing represents a value that could not be obtained from the cluster
+	templateMissing = "missing"
+
+	// catalogImageTemplateAnnotation is OLM annotation. The annotation value that corresponds
+	// to this key is used as means to adjust a catalog source image, where templated
+	// values are replaced with values found in the cluster
+	CatalogImageTemplateAnnotation = "olm.catalogImageTemplate"
+)
+
+// templateNameToReplacementValuesMap is storage for templates and their resolved values
+// The values are initialized to variable "templateMissing"
+var templateNameToReplacementValuesMap = map[string]string{}
+
+// convertToKey is a function that creates a key for templateNameToReplacementValuesMap based on a GVK key and json path
+func convertToKey(key GVK_Key, jsonPath string) string {
+	return fmt.Sprintf("{group:%s,version:%s,kind:%s,name:%s,namespace:%s,jsonpath:%s}", key.Group, key.Version, key.Kind, key.name, key.namespace, jsonPath)
+}
+
+// gvkToJSONPathMap is a multimap (i.e. one key with multiple values) where each value is
+// zero or more JSON paths. In other words the user could specify multiple JSON path references
+// for the same kubernetes manifest
+var gvkToJSONPathMap = map[GVK_Key][]string{}
+
+func init() {
+	// Handle known static templates
+	initializeIfNeeded(TEMPL_KUBEMAJORV)
+	initializeIfNeeded(TEMPL_KUBEMINORV)
+	initializeIfNeeded(TEMPL_KUBEPATCHV)
+}
+
+// initializeIfNeeded sets the map to a "missing" value if its not already present
+func initializeIfNeeded(templateKey string) {
+	// have we encountered this template already?
+	if _, ok := templateNameToReplacementValuesMap[templateKey]; !ok {
+		// this is a new template, so default to missing value
+		templateNameToReplacementValuesMap[templateKey] = templateMissing
+	}
+}
+
+// resetMaps is only useful for test cases
+func resetMaps() {
+	templateNameToReplacementValuesMap = map[string]string{}
+	gvkToJSONPathMap = map[GVK_Key][]string{}
+	initializeIfNeeded(TEMPL_KUBEMAJORV)
+	initializeIfNeeded(TEMPL_KUBEMINORV)
+	initializeIfNeeded(TEMPL_KUBEPATCHV)
+}
+
+type regexEntry struct {
+	captureGroup string
+	template     string
+}
+
+func (r *regexEntry) String() string {
+	return fmt.Sprintf(`(?P<%s>%s)`, r.captureGroup, r.template)
+}
+
+type regexEntries []regexEntry
+
+func (r regexEntries) String() string {
+	regexEntryAsString := []string{}
+	for _, regexEntry := range r {
+		regexEntryAsString = append(regexEntryAsString, regexEntry.String())
+	}
+	result := strings.Join(regexEntryAsString, "|")
+	return result
+}
+
+var regexList = regexEntries{
+	{cap_grp_kubemajorv, TEMPL_KUBEMAJORV},
+	{cap_grp_kubeminorv, TEMPL_KUBEMINORV},
+	{cap_grp_kubepatchv, TEMPL_KUBEPATCHV},
+	{cap_grp_gvk, templ_gvk},
```

Comment:
do we need to define or compile these if they're not going to be used in this PR? I'd like to cut code for features that are in question.

Reply:
fixed